keps/sig-storage/1698-generic-ephemeral-volumes/README.md

```diff
@@ -0,0 +1,566 @@
+Memory](https://memcached.org/blog/persistent-memory/) (PMEM) instead
+of normal DRAM. When deploying memcached through one of the app
+controllers, `EphemeralVolumeSource` makes it possible to request a volume
+of a certain size from a CSI driver like
+[PMEM-CSI](https://github.com/intel/pmem-csi).
+
+#### Local LVM storage as scratch space
+
+Applications working with data sets that exceed the RAM size can
+request local storage with performance characteristics or size that is
+not met by the normal Kubernetes `EmptyDir` volumes. For example,
+[TopoLVM](https://github.com/cybozu-go/topolvm) was written for that
+purpose.
+
+#### Read-only access to volumes with data
+
+Provisioning a volume might result in a non-empty volume:
+- [restore a snapshot](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#volume-snapshot-and-restore-volume-from-snapshot-support)
+- [cloning a volume](https://kubernetes.io/docs/concepts/storage/volume-pvc-datasource)
+- [generic data populators](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20200120-generic-data-populators.md)
+
+For those, it might make sense to mount the volume read-only inside
+the pod to prevent accidental modification of the data. For example,
+the goal might be to just retrieve the data and/or copy it elsewhere.
+
+### Risks and Mitigations
+
+Enabling this feature allows users to create PVCs indirectly if they can
+create pods, even if they do not have permission to create them
+directly. Cluster administrators must be made aware of this. If this
+does not fit their security model, they can disable the feature
+through the feature gate that will be added for the feature.
+
+In addition, with a new
+[`FSType`](https://github.com/kubernetes/kubernetes/blob/1fb0dd4ec5134014e466509163152112626d52c3/pkg/apis/policy/types.go#L278-L309)
+it will be possible to limit the usage of this volume source via the
+[PodSecurityPolicy
+(PSP)](https://kubernetes.io/docs/concepts/policy/pod-security-policy/#volumes-and-file-systems).
+
+The normal namespace quota for PVCs in a namespace still applies, so
+even if users are allowed to use this new mechanism, they cannot use
+it to circumvent other policies.
+
+## Example
+
+Here is a full example for a higher-level object that uses a generic ephemeral
+inline volume:
+
+```
+apiVersion: apps/v1
+kind: DaemonSet
+metadata:
+  name: fluentd-elasticsearch
+  namespace: kube-system
+spec:
+  selector:
+    matchLabels:
+      name: fluentd-elasticsearch
+  template:
+    metadata:
+      labels:
+        name: fluentd-elasticsearch
+    spec:
+      containers:
+      - name: fluentd-elasticsearch
+        image: quay.io/fluentd_elasticsearch/fluentd:v2.5.2
+        volumeMounts:
+        - name: varlog
+          mountPath: /var/log
+        - name: scratch
+          mountPath: /scratch
+      volumes:
+      - name: varlog
+        hostPath:
+          path: /var/log
+      - name: scratch
+        ephemeral:
```

Comment:
In general, this all looks good.

The only thing I'm not sure of is the naming here.

I would have loved to have the volume type ephemeral for the local inline ephemeral use case we have now. But I was told it had to hang off of the csi driver. 

Having some ephemeral be under csi and some under volumeSource would be hard to explain to users I think.

At some point, maybe pod api v2 whenever that happens, the csi driver should probably fold back out so it looks like its more at the root of the volumeSource struct. As all drivers but configmap/secret/emptydir/hostpath will be moving to csi, so having csi in the api kind of stops making sense at some point.

All in all, I'm not against this. I think in the long run its more usable. But, I think in the intermediate time its kind of confusing to end users. Perhaps we should design the api here such that it could be used for all kinds of ephemeral storage. And we treat the csi version as an alias and deprecate it? In my mind, the local ephemeral drivers are just an optimization (shortcuts) of the usage of general ephemeral inline. So to a user, ideally should look the same. They shouldn't have to care they are different.

Reply:
> But, I think in the intermediate time its kind of confusing to end users. Perhaps we should design the api here such that it could be used for all kinds of ephemeral storage. And we treat the csi version as an alias and deprecate it?

This is part of the long term plan and the reason why `VolumeClaimTemplate` is a pointer: we can add a new, alternative pointer to a struct with the values for a "CSI ephemeral inline volume", allow that as alternative and then with a suitable deprecation period, remove the `CSI` entry in `VolumeSource` in favor of this unified `Ephemeral` entry.

Reply:
I can see how that might work with the raw api. but what about the yaml representation? How will it know what type of object to generate?

Reply:
Ya we might need a wrapper struct here `PersistentVolumeClaimTemplate` to disambiguate and be consistent with existing `PodTemplate`.